### PR TITLE
Normalize functions

### PR DESCRIPTION
**Pull request by jorgefandinno**
This pull request contains a new function to normalize clingo.Function as ast.Function.

### CODE REVIEW

**Comment by rkaminsk** [comment]
Shouldn't the transformer also transform the arguments?

**Comment by jorgefandinno** [comment]
> Shouldn't the transformer also transform the arguments?

Is it possible to have an ast.Function as an argument to a SymbolType.Function? I understood that this was not the case from the API.

**Comment by rkaminsk** [comment]
> > Shouldn't the transformer also transform the arguments?
> 
> Is it possible to have an ast.Function as an argument to a SymbolType.Function? I understood that this was not the case from the API.

I might be confusing things but you are using `ast.Function` and passing a list of symbols as argument.

Instead of
```python
        arguments = symbol.arguments

        return ast.Function(location, name, arguments, False)
```
I would have expected something like
```python
        arguments = [symbol_to_ast(arg) for arg in symbol.arguments]

        return ast.Function(location, name, arguments, False)
```

File: clingox/ast.py [hunk 1]
```diff
@@ -1434,6 +1459,6 @@ def visit_SymbolicTerm(self, term: AST):  # pylint: disable=invalid-name
 
         location = term.location
         name = symbol.name
-        arguments = symbol.arguments
+        arguments = [symbol_to_ast(arg, location) for arg in symbol.arguments]
 
-        return ast.Function(location, name, arguments, False)
+        return ast.Function(location, name, arguments, external=False)
```
Comment (rkaminsk) [inline clingox/ast.py:1464]: I think you can simplify this calling `symbol_to_ast(location, symbol)`.

File: clingox/ast.py [hunk 2]
```diff
@@ -1391,3 +1391,70 @@ def theory_term_to_literal(x: AST, parse: bool = True) -> AST:
     if parse:
         x = clingo_literal_parser()(x)
     return _theory_term_to_literal(x, True, ast.Sign.NoSign)
+
+
+def normalize_symbolic_terms(x: AST):
```
Comment (rkaminsk) [inline clingox/ast.py:1396]: Please add this to ALL.

File: clingox/ast.py [hunk 3]
```diff
@@ -1391,3 +1391,70 @@ def theory_term_to_literal(x: AST, parse: bool = True) -> AST:
...
+        The AST to rewrite.
+
+    Returns
+    -------
+    The rewritten AST.
+    """
+    return _NormalizeSymbolicTermTransformer().visit(x)
+
+
+def symbol_to_ast(x: clingo.Symbol, location: ast.Location) -> AST:
```
Comment (rkaminsk) [inline clingox/ast.py:1413]: I think we can make this a private function prefixing it with underscore. At least I can't imagine that this is something that would often be used in practice.

File: clingox/ast.py [hunk 4]
```diff
@@ -1391,3 +1391,70 @@ def theory_term_to_literal(x: AST, parse: bool = True) -> AST:
...
+class _NormalizeSymbolicTermTransformer(Transformer):
+    """Transforms a SymbolicTerm AST of type Function into an AST of type ast.Function."""
+
+    def visit_SymbolicTerm(self, term: AST):  # pylint: disable=invalid-name
+        """
+        Transform the given symbolic term.
+
+        Parameters
+        ----------
+        x
```
Comment (rkaminsk) [inline clingox/ast.py:1447]: Does not match the parameter name.

**Review by rkaminsk — CHANGES_REQUESTED** [review]

**Pull request merged**